R/monitor.R

```diff
@@ -123,17 +87,89 @@ vetiver_compute_metrics <- function(data,
+#' overlapping dates with the existing pin.If `TRUE`, overwrite any metrics for
+#' dates that exist both in the existing pin and new metrics with the _new_
+#' values.
+#'
+#' @return A dataframe of metrics.
+#'
+#' @details Sometimes when you monitor a model at a given time aggregation, you
+#' may end up with dates in your new metrics (like `new_metrics` in the example)
+#' that are the same as dates in your existing aggregated metrics (like
+#' `original_metrics` in the example). This can happen if you need to re-run a
+#' monitoring report because something failed. With `overwrite = FALSE` (the
+#' default), `vetiver_pin_metrics()` will error when there are overlapping
+#' dates. With `overwrite = TRUE`, `vetiver_pin_metrics()` will replace such
+#' metrics with the new values. You probably want `FALSE` for interactive use
+#' and `TRUE` for dashboards or reports that run on a schedule.
+#'
+#' You can initially create your pin with `type = "arrow"` or the default
+#' (`type = "rds"`). `vetiver_pin_metrics()` will update the pin using the
+#' same `type`.
+#'
+#' @examplesIf rlang::is_installed(c("dplyr", "parsnip", "modeldata"))
+#' library(dplyr)
+#' library(parsnip)
+#' data(Chicago, package = "modeldata")
+#' Chicago <- Chicago %>% select(ridership, date, all_of(stations))
+#' training_data <- Chicago %>% filter(date < "2009-01-01")
+#' testing_data <- Chicago %>% filter(date >= "2009-01-01", date < "2011-01-01")
+#' monitoring <- Chicago %>% filter(date >= "2011-01-01", date < "2012-12-31")
+#' lm_fit <- linear_reg() %>% fit(ridership ~ ., data = training_data)
+#'
+#' library(pins)
+#' b <- board_temp()
+#'
+#' ## before starting monitoring, initiate the metrics and pin
+#' ## (for example, with the testing data):
+#' original_metrics <-
+#'     augment(lm_fit, new_data = testing_data) %>%
+#'     vetiver_compute_metrics(date, "week", ridership, .pred, every = 4L)
+#' pin_write(b, original_metrics, "lm_fit_metrics", type = "arrow")
+#'
+#' ## to continue monitoring with new data, compute metrics and update pin:
+#' new_metrics <-
+#'     augment(lm_fit, new_data = monitoring) %>%
+#'     vetiver_compute_metrics(date, "week", ridership, .pred, every = 4L)
+#' vetiver_pin_metrics(b, new_metrics, "lm_fit_metrics")
+#'
+#' @seealso [vetiver_compute_metrics()], [vetiver_plot_metrics()]
 #' @export
 vetiver_pin_metrics <- function(board,
                                 df_metrics,
                                 metrics_pin_name,
                                 .index = .index,
-                                overwrite = FALSE) {
+                                overwrite = FALSE,
+                                ...) {
     .index <- enquo(.index)
     .index <- eval_select_one(.index, df_metrics, "date_var")
     new_dates <- unique(df_metrics[[.index]])
 
+    old_type <- pins::pin_meta(board, metrics_pin_name)$type
+    if (! old_type %in% c("arrow", "rds")) {
```

Comment:
What's the reason CSV is not supported? Is it because `pin_read()'ing` a CSV won't read in date columns as dates?

(I think a similar challenge occurred on the python side, but I had it attempt to coerce the date index to a date for better or worse 😬, so might be good to align behavior one way or the other..!)

Reply:
I outlined this a bit over in #118 but in R, it's not only the date but also the metrics columns and such (pins uses the base R `read.csv()` which is famously troublesome). There would be quite a bit of coercing and type fixing to do with CSV, so much so that I think enforcing something like arrow is a good move. I am open to users bringing this up later again, though, if CSV is important for some situations.

Reply:
That makes sense! I noticed the presence of the infamous strings are being read as factors situation while checking pin_read w/ CSV 😅

Reply:
I've added #123 to track interest in CSV metrics storage for the R side 👍